episodes/12-fire-forget-tasks.md

```diff
@@ -0,0 +1,392 @@
+their messages in the corresponding position. Then, at the end, have the main task printing all elements of
+the array in order.
+
+:::::::::::::::::::::::: solution
+
+The following code is a possible solution:
+
+```chpl
+var x=1;
+config var numoftasks=2;
+var messages: [1..numoftasks] string;
+
+writeln("This is the main task: x = ",x);
+
+coforall taskid in 1..numoftasks do
+{
+   var c=taskid+1;
+   var s="this is task "+taskid+": x + "+taskid+" = "+(x+taskid)+". My value of c is: "+c;
+   messages[taskid]=s;
+}
+
+for i in 1..numoftasks do writeln(messages[i]);
+writeln("this message won't appear until all tasks are done...");
+```
+
+```bash
+chpl exercise_coforall.chpl -o exercise_coforall
+./exercise_coforall --numoftasks=5
+```
+
+```output
+This is the main task: x = 1
+this is task 1: x + 1 = 2. My value of c is: 2
+this is task 2: x + 2 = 3. My value of c is: 3
+this is task 3: x + 3 = 4. My value of c is: 4
+this is task 4: x + 4 = 5. My value of c is: 5
+this is task 5: x + 5 = 6. My value of c is: 6
+this message won't appear until all tasks are done...
+```
+
+Note that `+` is a **_polymorphic_** operand in Chapel. In this case it concatenates `strings` with `integers`
+(which are transformed to strings).
+
+:::::::::::::::::::::::::::::::::
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: challenge
+
+## Challenge 2: Can you do it?
+
+Consider the following code:
+
+```chpl
+use Random;
+config const nelem=5000000000;
+var x: [1..n] int;
+fillRandom(x);	//fill array with random numbers
+var mymax=0;
+
+// here put your code to find mymax
+
+writeln("the maximum value in x is: ",mymax);
+```
+
+Write a parallel code to find the maximum value in the array x.
+
+:::::::::::::::::::::::: solution
+
+```chpl
+config const numoftasks=12;
+const n=nelem/numoftasks;
+const r=nelem-n*numoftasks;
+
+var d: [0..numoftasks-1] real;
+
+coforall taskid in 0..numoftasks-1 do
+{
+  var i: int;
+  var f: int;
+  if taskid<r then
+  {
+     i=taskid*n+1+taskid;
+     f=taskid*n+n+taskid+1;
+  }
+ else
+ {
+      i=taskid*n+1+r;
+      f=taskid*n+n+r;
+  }
+  for c in i..f do
+  {
+      if x[c]>d[taskid] then d[taskid]=x[c];
+  }
+}
+for i in 0..numoftasks-1 do
+{
+  if d[i]>mymax then mymax=d[i];
+}
+```
+
+```bash
+> > chpl --fast exercise_coforall_2.chpl -o exercise_coforall_2
+> > ./exercise_coforall_2
+```
+
+```output
+the maximum value in x is: 1.0
+```
+
+We use the coforall to spawn tasks that work concurrently in a fraction of the array. The trick here is to
+determine, based on the _taskid_, the initial and final indices that the task will use. Each task obtains the
+maximum in its fraction of the array, and finally, after the coforall is done, the main task obtains the
+maximum of the array from the maximums of all tasks.
+
+:::::::::::::::::::::::::::::::::
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Discussion
```

Comment:
Specify the numbers of tasks and array sizes learners should use to see the intended effect within a reasonable amount of time. Ambiguous instructions like this place high load on the Instructor.

Reply:
I fixed the solution in this exercise: someone modified it years ago, but in such a way that the solution was too complex and did not even compile. I removed this discussion on playing with the array size and the number of tasks. The original intent of this discussion was to confirm that the problem stops scaling when the number of tasks exceeds the available number of cores, but this example is really bad for this: it is too fast to compute (so the small wallclock time is not reliable), and asking for nelem > 1e9 will use all your memory, so that on a cluster the job will simply crash.